// @sogni-ai/sogni-client/dist/Account/CurrentAccount.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const DataEntity_1 = __importDefault(require("../lib/DataEntity"));
function getDefaults() {
    return {
        networkStatus: 'disconnected',
        network: null,
        balance: {
            sogni: {
                credit: '0',
                debit: '0',
                net: '0',
                settled: '0'
            },
            spark: {
                credit: '0',
                debit: '0',
                net: '0',
                settled: '0',
                premiumCredit: '0'
            }
        },
        walletAddress: undefined,
        username: undefined,
        token: undefined,
        refreshToken: undefined
    };
}
/**
 * Current account data.
 * @expand
 */
class CurrentAccount extends DataEntity_1.default {
    constructor(data) {
        super(data || getDefaults());
    }
    _clear() {
        this._update(getDefaults());
    }
    get isAuthenicated() {
        return !!this.data.refreshToken;
    }
    get networkStatus() {
        return this.data.networkStatus;
    }
    get network() {
        return this.data.network;
    }
    get balance() {
        return this.data.balance;
    }
    get walletAddress() {
        return this.data.walletAddress;
    }
    get username() {
        return this.data.username;
    }
    get token() {
        return this.data.token;
    }
    get refreshToken() {
        return this.data.refreshToken;
    }
}
exports.default = CurrentAccount;
